packages/adapters/eth/src/failover/client.rs

```diff
@@ -0,0 +1,1023 @@
+            shared_state: Arc::new(Mutex::new(shared_state)),
+            provider_constants: ProviderConstants {
+                commit_interval,
+                contract_caller_address,
+                blob_poster_address,
+            },
+            metrics: Metrics::default(),
+        })
+    }
+
+    pub fn health_checker(&self) -> HealthChecker
+    where
+        I::Provider: 'static,
+    {
+        Box::new(HealthReporter {
+            shared_state: Arc::clone(&self.shared_state),
+            health_thresholds: self.health_thresholds,
+        })
+    }
+
+    /// Retrieves the currently active provider if it's healthy, or tries to fail over to the next.
+    async fn get_healthy_provider(&self) -> EthResult<ProviderHandle<I::Provider>> {
+        let mut state = self.shared_state.lock().await;
+
+        if state
+            .active_provider
+            .is_healthy(&self.health_thresholds)
+            .await
+        {
+            debug!(
+                "Using current provider '{}' which is healthy",
+                state.active_provider.name
+            );
+            return Ok(state.active_provider.clone());
+        }
+
+        info!(
+            "Current provider '{}' is unhealthy, attempting to fail over to next provider",
+            state.active_provider.name
+        );
+
+        let handle =
+            connect_to_first_available_provider(&self.initializer, &mut state.remaining_endpoints)
+                .await?;
+
+        info!("Successfully failed over to provider '{}'", handle.name);
+
+        state.active_provider = handle.clone();
+        Ok(handle)
+    }
+
+    /// Note a mempool drop for the current provider
+    pub async fn note_mempool_drop(&self, reason: &str) -> EthResult<()> {
+        // Get the current provider without checking health since we don't want a failover here
+        let provider = self.shared_state.lock().await.active_provider.clone();
+        info!(
+            "Noting mempool drop on provider '{}': {reason}",
+            provider.name
+        );
+
+        provider
+            .note_mempool_drop(reason, self.health_thresholds.mempool_drop_window)
+            .await;
+
+        self.metrics
+            .eth_mempool_drops
+            .with_label_values(&[&provider.name])
+            .inc();
+
+        Ok(())
+    }
+
+    /// Core abstraction to execute operations with failover logic
+    async fn execute_operation<F, Fut, T>(&self, operation_factory: F) -> EthResult<T>
+    where
+        F: FnOnce(Arc<I::Provider>) -> Fut + Send + Sync,
+        Fut: std::future::Future<Output = EthResult<T>> + Send,
+        T: Send,
+    {
+        let provider = self.get_healthy_provider().await?;
+        let provider_name = provider.name.as_str();
+
+        debug!("Executing operation on provider '{}'", provider_name);
+
+        let result = operation_factory(Arc::clone(&provider.provider)).await;
+
+        match result {
+            Ok(value) => {
```

Comment:
If `operation_factory` was a factory, I'd expect the output of the function to be an operation, not the result of calling an operation. It seems like this is just the operation, and hence I think we should rename it.

```suggestion
    async fn execute_operation<F, Fut, T>(&self, operation: F) -> EthResult<T>
    where
        F: FnOnce(Arc<I::Provider>) -> Fut + Send + Sync,
        Fut: std::future::Future<Output = EthResult<T>> + Send,
        T: Send,
    {
        let provider = self.get_healthy_provider().await?;
        let provider_name = provider.name.as_str();

        debug!("Executing operation on provider '{}'", provider_name);

        let result = operation(Arc::clone(&provider.provider)).await;

        match result {
            Ok(value) => {
```